### PR TITLE
Instrument dictation tap-stall regression with diagnostic logging

### PR DESCRIPTION
**Pull request by moona3k**
## Summary

Intermittent silent tap stall in dictation: press Fn, talk for 10–18 s, release, see "more audio please" because zero buffers arrived. Reproduced 2026-05-03 — file `664D3EA7-…wav`, 18.3 s elapsed, `input_buffers=0` — while captures bracketing it on the same code path worked fine. Almost certainly a regression introduced by PR #189 (shared-mic-engine, Apr 30). The existing log goes silent between `engine_started` and `stop`, so we can't yet tell macOS-side HAL rearbitration apart from a teardown/start window we opened.

**Log-only diagnostic package. No UX change. No recording aborts.** A "microphone isn't responding" toast would mask the regression as a fact of life; the goal is to find what changed and fix *that*.

## What the log looks like now

Three signatures, all unambiguous:

```
HEALTHY (silent user, thinking before speaking)
─────────────────────────────────────────────────────────────────
engine_started file=…  default_input=89:MacBook Pro Microphone  sr=48000  ch=1
first_buffer                                                  ← +108 ms
heartbeat  input_buffers=300  isRunning=true                  ← +5 s
... user thinks → speaks → releases Fn ...
stop  sample_count=N  input_buffers=N  non_silent_buffers=M

STALL (the bug we're hunting)
─────────────────────────────────────────────────────────────────
engine_started file=…  default_input=89:MacBook Pro Microphone  sr=48000  ch=1
              ← no first_buffer ever fires
no_buffers_within_timeout  isRunning=true  default_input=…    ← +2 s
heartbeat  input_buffers=0  isRunning=true                    ← +5 s
heartbeat  input_buffers=0  isRunning=true                    ← +10 s
stop  sample_count=0  input_buffers=0

STALL CORRELATED WITH OS EVENT (cause: macOS,  fix: reactive)
─────────────────────────────────────────────────────────────────
engine_started …
shared_mic_engine_configuration_changed  sr=48000  ch=1  isRunning=true
no_buffers_within_timeout  isRunning=true  …
```

The heartbeat measures **buffer delivery**, not speech — silent users still receive buffers (full of zeros), so they're never conflated with the stall.

## New events (all in `~/Library/Logs/MacParakeet/dictation-audio.log`)

- `dictation_diagnostics_session_start pid=… version=… build=… commit=…` — once at app launch; splits dev-app and `swift test` runs in the shared on-disk log
- `dictation_capture_engine_started … default_input=… sr=… ch=…` — enriched with device + format
- `dictation_capture_no_buffers_within_timeout isRunning=… default_input=…` — fires 2 s after `engine_started` if no buffer arrived; recording continues
- `dictation_capture_heartbeat input_buffers=N input_frames=M isRunning=… default_input=…` — every 5 s while recording
- `dictation_capture_start_cancelled_during_subscribe` — `stop()` raced into `start()` mid-await
- `shared_mic_diagnostics transition=… subscribers=… vpio_engaged=… …` — on every state-machine transition
- `shared_mic_engine_configuration_changed sr=… ch=… isRunning=… default_input=…` — Core Audio renegotiated the input chain (most likely external trigger)
- Mirrored all remaining os.log-only platform events (`shared_mic_engine_stopped`, `_input_device_started`, `_set_failed`, `_start_failed`, `_input_format_failed`, `_reconfigure_failed`, `_stop_failed`) so the on-disk log is self-contained

## Decision rule for the next captured stall

```
       no_buffers_within_timeout
                  │
   ┌──────────────┴──────────────┐
   │                             │
preceded by                  preceded by
configuration_changed?       no signal at all?
   │                             │
   ▼                             ▼
macOS-side cause           our shared-engine
fix: reactive re-tap on    teardown/start window
configuration change       suspect: PR #189
                           in startEngineLocked
```

## Working hypothesis (lead, not proof)

**HAL deactivation during idle, racy wake-up on the freshly-created `AVAudioEngine`.** Pre-#189 dictation owned a long-lived engine; post-#189 the platform tears down and recreates `AVAudioEngine()` on every full unsubscribe. With nothing in our process holding the input device, macOS power-management deactivates the HAL after a short idle. The next `audioEngine.start()` returns success before the HAL has fully re-bound the input chain — engine reports `isRunning=true`, tap is installed against a node with no upstream audio source, no buffers flow.

Confidence: ~60–70 %. The new logging will resolve this — heartbeat showing `input_buffers=0 isRunning=true` plus the **absence** of a `shared_mic_engine_configuration_changed` event during the stall window is strong evidence for the hunch. Two competing hypotheses (residual VPAU state, configuration race) would each leave their own signature; the diagnostic package is built to distinguish them.

**Trigger conditions if the hunch is right.** All three required:
1. Previous dictation/meeting fully ended → no subscribers → engine torn down.
2. No other process holding the mic (no Zoom, Discord, Voice Memos, other transcription apps).
3. Enough idle time for the HAL to deactivate (empirically minutes; probably tens of seconds is enough).

Risk multipliers: battery / Low Power Mode, external Bluetooth/USB mics, just woke from sleep, newer macOS releases. Heavy dictators are basically immune (HAL stays warm). Sporadic dictators on battery with AirPods are the most exposed profile.

**Likely fix when the next stall confirms the hypothesis:** keep the shared engine alive once started, only stopping at app quit. Reverts the lifecycle behavior to what worked pre-#189. Marginal cost: trace mic activity when no one's using dictation. Marginal benefit: no HAL cold-start race.

## Where the log lives & how users share it

`~/Library/Logs/MacParakeet/dictation-audio.log` — same path on every build flavor (dev binary, DMG, eventual TestFlight). Written via `AudioCaptureDiagnostics.append` → `AppPaths.logsDir`. Local only; never auto-uploaded. Telemetry is a separate channel (ADR-012, anonymous counters only, no log content).

When a user reports the stall, ask them to share via either:
- Finder ⌘⇧G → paste `~/Library/Logs/MacParakeet/` → drag `dictation-audio.log` into the issue
- Terminal: `cat ~/Library/Logs/MacParakeet/dictation-audio.log | pbcopy` → paste

Privacy-safe: timestamps, device names, counters. **No audio, no transcripts.** Same content shape that already lives in feedback issues.

**Retention.** `maxLogBytes` is **5 MB** (bumped from 1 MB in this PR — the new heartbeat roughly doubled per-recording log volume). When the file crosses 5 MB it's **deleted** (not rotated). At the new event volume that's tens of days of cumulative recording for a heavy user — enough that a stall reported via the feedback flow still has its surrounding ~30 s context window intact.

## Follow-ups (not in this PR)

- **`dictation.silent_stall` telemetry counter** — fire from the watchdog. One event name, no payload beyond what telemetry already includes. Gives field frequency without exposing log content; pairs with the local log: counter says *how often*, log (when shared) says *why*.
- **Log-snippet capture in the feedback form** — when "audio problem" is picked from the category dropdown, attach the last N lines automatically. Turns every reported stall into actionable data without users needing to know about the log path.

## Why no tests

`MicrophoneCapture.scheduleSilentBufferWatchdog` (the reference watchdog this mirrors) has zero tests for the same reasons: timing-sensitive, observability-only, AVAudioEngine state is hard to fake without flake risk. Adding flaky tests for a log line would be net-negative.

## Test plan

- [x] `swift build --target MacParakeetCore` — clean
- [x] `swift test` — 2205 tests pass, 1 skipped, 0 failures (post-rebase onto `bd62e148`)
- [ ] Dev-app smoke: launch via `scripts/dev/run_app.sh`, dictate a few times, confirm `dictation_diagnostics_session_start` and `heartbeat` lines land at sensible cadence
- [ ] Reviewer: confirm `armCaptureDiagnostics` closures bail via the `sessionGeneration` guard when `stop()` raced ahead
- [ ] Reviewer: confirm the `AVAudioEngineConfigurationChangeNotification` observer is removed in **every** engine teardown path (`tearDownLocked`, `resetEngineLocked`, `replaceEngineAfterFailureLocked`)

## Background

Full diagnosis, regression hypothesis, surrounding-session evidence, and historical log excerpts: `journal/2026-05-03-dictation-silent-stall.md` (private journal, not in this PR).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: Sources/MacParakeetCore/Audio/MicrophoneEnginePlatform.swift [hunk 1]
```diff
@@ -293,13 +318,62 @@ public final class AVAudioEngineMicrophonePlatform: MicrophoneEnginePlatform, @u
...
+                    isRunning: self.running
+                )
+            }
+            AudioCaptureDiagnostics.append(
+                "shared_mic_engine_configuration_changed sr=\(snapshot.sr) ch=\(snapshot.ch) isRunning=\(snapshot.isRunning) default_input=\(AudioCaptureDiagnostics.defaultInputDeviceLabel())"
+            )
+            self.logger.info(
+                "shared_mic_engine_configuration_changed sr=\(snapshot.sr, privacy: .public) ch=\(snapshot.ch, privacy: .public) isRunning=\(snapshot.isRunning, privacy: .public)"
+            )
+        }
```
[the remark this answers is not in the document]
Comment (moona3k) [reply Sources/MacParakeetCore/Audio/MicrophoneEnginePlatform.swift:363]: Fixed in `92fe7f33`. The observer now captures `notification.object as? AVAudioEngine`, dispatches asynchronously onto the platform queue, and reads the format from that captured engine instead of `self.audioEngine`, so the notification callback no longer synchronously re-enters `self.queue` or races the mutable engine property.

I also verified the PR branch and a synthetic merge into current `origin/main` with `swift build --target MacParakeetCore` and `swift test`.

**Comment by moona3k** [comment]
Final review: LGTM.

- Addressed Gemini's high-priority `AVAudioEngineConfigurationChange` observer comment in `92fe7f33` and replied inline.
- CodeRabbit's remaining thread is an informational rate-limit notice, no code action required.
- Verified the PR branch and a synthetic merge into current `origin/main`:
  - `swift build --target MacParakeetCore`
  - `swift test` — 2205 tests, 1 skipped, 0 failures

No remaining blockers from my review.
